src/Engine.php

```diff
@@ -0,0 +1,201 @@
+    {
+        $this->client = $client;
+    }
+
+    /**
+     * Save entities to Algolia.
+     *
+     * @param array<int, SearchableEntity>    $searchableEntities
+     * @param array<string, int|string|array> $requestOptions
+     *
+     * @return array<string, BatchIndexingResponse>
+     *
+     * @throws \Algolia\AlgoliaSearch\Exceptions\AlgoliaException
+     */
+    public function index($searchableEntities, $requestOptions = [])
+    {
+        if ($searchableEntities instanceof SearchableEntity) {
+            $searchableEntities = [$searchableEntities];
+        }
+
+        $data = [];
+        foreach ($searchableEntities as $entity) {
+            $searchableArray = $entity->getSearchableArray();
+            if ($searchableArray === null || count($searchableArray) === 0) {
+                continue;
+            }
+
+            $indexName = $entity->getIndexName();
+
+            if (!isset($data[$indexName])) {
+                $data[$indexName] = [];
+            }
+
+            $data[$indexName][] = $searchableArray + [
+                    'objectID' => $entity->getId(),
+                ];
+        }
+
+        $result = [];
+        foreach ($data as $indexName => $objects) {
+            if (!array_key_exists('autoGenerateObjectIDIfNotExist', $requestOptions)) {
+                $requestOptions['autoGenerateObjectIDIfNotExist'] = true;
+            }
+
+            $result[$indexName] = $this->client
+                ->initIndex($indexName)
+                ->saveObjects($objects, $requestOptions);
+        }
+
+        return $result;
+    }
+
+    /**
+     * Remove entities from Algolia.
+     *
+     * @param array<int, SearchableEntity>    $searchableEntities
+     * @param array<string, int|string|array> $requestOptions
+     *
+     * @return array<string, BatchIndexingResponse>
+     *
+     * @throws \Algolia\AlgoliaSearch\Exceptions\AlgoliaException
+     */
+    public function remove($searchableEntities, $requestOptions = [])
+    {
+        if ($searchableEntities instanceof SearchableEntity) {
+            $searchableEntities = [$searchableEntities];
+        }
+
+        $data = [];
+        foreach ($searchableEntities as $entity) {
+            $searchableArray = $entity->getSearchableArray();
+            if ($searchableArray === null || count($searchableArray) === 0) {
+                continue;
+            }
+            $indexName = $entity->getIndexName();
+
+            if (!isset($data[$indexName])) {
+                $data[$indexName] = [];
+            }
+
+            $data[$indexName][] = $entity->getId();
+        }
+
+        $result = [];
+        foreach ($data as $indexName => $objects) {
+            $result[$indexName] = $this->client
+                ->initIndex($indexName)
+                ->deleteObjects($objects, $requestOptions);
+        }
+
+        return $result;
+    }
+
+    /**
+     * Clear all objects from the index.
```

Comment:
nit: What I've done for the C#/Java client is to take methods description from the documentation which are very well written. For example the
[clear method](https://www.algolia.com/doc/api-reference/api-methods/clear-objects/)

```
Clear the records of an index without affecting its settings.

This method enables you to delete an index’s contents (records) without removing any settings, rules and synonyms.

If you want to remove the entire index and not just its records, use the delete method instead.
```